### PR TITLE
[engine] stack-safe kind-checking

### PR DESCRIPTION
**Pull request by nickchapman-da**
Make kind-checking be stack-safe:
- [x] add new test-cases to demonstrate stack-overflow
- [x] fix stack-overflow

This advances #13410

### CODE REVIEW

**Review by carlpulley-da — APPROVED** [review]
LGTM

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 1]
```diff
@@ -460,9 +471,8 @@ private[validation] object Typing {
       case _ => typ0
     }
 
-    private[Typing] def checkRecordType(fields: ImmArray[(FieldName, Type)]): Unit = {
-      checkUniq[FieldName](fields.keys, EDuplicateField(ctx, _))
-      fields.values.foreach(checkType(_, KStar))
+    private[Typing] def checkRecordTypeTop(fields: ImmArray[(FieldName, Type)]): Unit = {
+      runWork(checkRecordType(fields) { Ret(()) })
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:476]: thank you
